scripts/helper/functions.sh

```diff
@@ -257,3 +257,31 @@ END
     exit 1
   fi
 }
+
+create_topic() {
+
+  local DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null && pwd )"
+
+  broker_host_port=$1
+  cluster_id=$2
+  topic_name=$3
+  confluent_value_schema_validation=$4
+  auth=$5
+
+  RESULT=$(curl -sS -X POST \
+    -u ${auth} \
+    --cacert /etc/kafka/secrets/snakeoil-ca-1.crt \
+    --header 'Content-Type: application/json' \
+    --header 'Accept: application/json' \
+    --data-binary @<(jq -n --arg topic_name "${topic_name}" --arg confluent_value_schema_validation "${confluent_value_schema_validation}" -f ${DIR}/topic.jq) \
+    "https://${broker_host_port}/kafka/v3/clusters/${cluster_id}/topics") && RC=$? || RC=$?
+
+  echo $RESULT | jq || true
+
+  if [[ $RC -ne 0 || -z $RESULT || $RESULT =~ "error_code" ]]; then
+    echo "ERROR: create topic failed $RESULT"
+    return 1
```

Comment:
Should this error code be handled by the calling script, i.e., fail fast?

Reply:
This should already fail-fast I think (and according to a manual test), via:

https://github.com/confluentinc/cp-demo/blob/d0649528f5d1f27e5b6f32ee682881369ac97ad0/scripts/helper/create-topics.sh#L2

If I change:

```
diff --git a/scripts/helper/create-topics.sh b/scripts/helper/create-topics.sh
index 7adfca4..d8700b1 100755
--- a/scripts/helper/create-topics.sh
+++ b/scripts/helper/create-topics.sh
@@ -10,6 +10,8 @@ echo "KAFKA_CLUSTER_ID: ${KAFKA_CLUSTER_ID}"

 auth="superUser:superUser"

+create_topic kafka1:8091 ${KAFKA_CLUSTER_ID} wikipedia.parsed.count-by-domain false ${auth}
+
 create_topic kafka1:8091 ${KAFKA_CLUSTER_ID} users true ${auth}
 create_topic kafka1:8091 ${KAFKA_CLUSTER_ID} wikipedia.parsed true ${auth}
 create_topic kafka1:8091 ${KAFKA_CLUSTER_ID} wikipedia.parsed.count-by-domain false ${auth}
```

... the script stops with:

```
ERROR: create topic failed {"error_code":40002,"message":"Topic 'wikipedia.parsed.count-by-domain' already exists."}
```

Is anything further required for fail-fast?

Reply:
@javabrett this test does not fail fast:

```
[11:53:59] ~/git/cp-demo(topic-creation-via-rest-v3) ✗: git diff
diff --git a/scripts/helper/create-topics.sh b/scripts/helper/create-topics.sh
index 7adfca4..64f386a 100755
--- a/scripts/helper/create-topics.sh
+++ b/scripts/helper/create-topics.sh
@@ -8,7 +8,7 @@ source ${DIR}/functions.sh
 KAFKA_CLUSTER_ID=$(get_kafka_cluster_id_from_container)
 echo "KAFKA_CLUSTER_ID: ${KAFKA_CLUSTER_ID}"
 
-auth="superUser:superUser"
+auth="superUser:superUser2"
 
 create_topic kafka1:8091 ${KAFKA_CLUSTER_ID} users true ${auth}
 create_topic kafka1:8091 ${KAFKA_CLUSTER_ID} wikipedia.parsed true ${auth}
```

`stdout` shows:

```
{
  "servlet": "default",
  "message": "Unauthorized",
  "url": "/kafka/v3/clusters/AO6RqB0HSnaagGflWwIy4w/topics",
  "status": "401"
}
```

But the start script continues and does not exit

Reply:
@ybyzek OK I see what happens there - for auth 401 errors, the error response is actually different, so we don't detect the error.

`curl` is well-known in terms of difficulty capturing all of good responses, error responses and HTTP codes - do we have an idiomatic way of doing that?

Reply:
I haven't played with this too much, but some suggestions to consider:

1. Parse output for "Unauthorized"
2. Check existence of topics afterwards

Reply:
@ybyzek I've pushed a commit which adds fetching of the HTTP response code from the `curl` command for `create_topic`, so we can test that in addition to the output scraping.  Unfortunately `curl` makes it fairly hard to fetch both output and response code, so the new bash script fragment is necessarily complicated.

The `create_topic` call now fails-fast for at least topic-already-exists (400) and unauthorized/bad-credentials (401).  Any response code >299 will be treated as fail.

Reply:
@javabrett beauty:

```
ERROR: create topic failed {
"servlet":"default",
"message":"Unauthorized",
"url":"/kafka/v3/clusters/QKMFRq6RSbeOfk4-6G7pPw/topics",
"status":"401"
}
```

It fails fast!!